apex - fused_rms_norm is broken

Running the tests `pytest thunder/tests/test_apex_fused_norms.py` gives following errors

<details>

```python
obj = <thunder.core.interpreter.WrappedValue object at 0x7507ec3d90c0>
args = (<thunder.core.interpreter.WrappedValue object at 0x7507ec3dacb0>, <thunder.core.interpreter.WrappedValue object at 0x...e.interpreter.WrappedValue object at 0x7507ec3da800>, <thunder.core.interpreter.WrappedValue object at 0x7507ec3d9bd0>), kwargs = {}
sequencify = <function sequencify at 0x750828d94ee0>
augmented_forward_impls = {<PrimIDs.ACOS: 68>: <function <lambda> at 0x750828a12830>, <PrimIDs.ACOSH: 69>: <function <lambda> at 0x750828a12f80>...mIDs.ASIN: 70>: <function <lambda> at 0x750828a13010>, <PrimIDs.ASINH: 71>: <function <lambda> at 0x750828a130a0>, ...}
backward_impls = {<PrimIDs.ACOS: 68>: <function <lambda> at 0x750828a13d00>, <PrimIDs.ACOSH: 69>: <function <lambda> at 0x750828a13d90>...mIDs.ASIN: 70>: <function <lambda> at 0x750828a13e20>, <PrimIDs.ASINH: 71>: <function <lambda> at 0x750828a13eb0>, ...}
jit_ctx = <thunder.core.jit_ext.JitCtx object at 0x7507ec3db610>, custom_fwd_bsyms = [t5 = thunder.executors.apex_fused_rms_norm_impl.fused_rms_norm(fc4, t_0, t_1, (3, 2), 1e-05, False)  # t5: "cuda:0 f32[4, 5, 3, 2]"]
orig_scopes = [[t5 = thunder.core.jit_ext.FusedRMSNormAffineMixedDtypesFunction(t_0, t_1, (3, 2), 1e-05, False)  # t5: "cuda:0 f32[4...xecutors.apex_fused_rms_norm_impl.fused_rms_norm(fc4, t_0, t_1, (3, 2), 1e-05, False)  # t5: "cuda:0 f32[4, 5, 3, 2]"]]
custom_autograd_function_cls = <class 'apex.normalization.fused_layer_norm.FusedRMSNormAffineMixedDtypesFunction'>, custom_forward = <function FusedRMSNormAffineMixedDtypesFunction.forward at 0x7508083c4d30>, ctx = <torch.autograd.function.FunctionCtx object at 0x7507ec3d9780>
wrapped_ctx = <thunder.core.interpreter.WrappedValue object at 0x7507ec3d9390>

    @register_general_jit_lookaside(torch.autograd.function.Function.apply.__func__)
    def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwargs):
        """Encapsulate forward into a bsym, define and register augmented fwd and bwd.
    
        This lookaside does three things:
            1. Encapsulate ``MyFunc(torch.autograd.Function).forward`` into :class:`~thunder.core.symbol.BoundSymbol`
            2. Define an augmented forward which is different from ``MyFunc.forward`` in its return values: returning
               ``ctx.saved_tensors`` as residuals.
            3. Trace ``MyFunc.backward``, define :class:`~thunder.core.trace.TraceCtx` whose args are ``(*residuals, *grads)``.
               So far, non-tensor ``ctx`` attributes seem to be folded into a trace.
        """
        from thunder.core.baseutils import check, sequencify
        from thunder.core.transforms import augmented_forward_impls, backward_impls, VJPDual
    
        jit_ctx: JitCtx = get_jit_ctx()
    
        custom_fwd_bsyms: list[BoundSymbol] = []
        orig_scopes = jit_ctx.computation_trace.scopes
    
        jit_ctx.computation_trace.scopes = [custom_fwd_bsyms]
    
        custom_autograd_function_cls = unwrap(obj)
        custom_forward = custom_autograd_function_cls.forward
        ctx = torch.autograd.function.FunctionCtx()
        ctx_proxy = proxy(ctx, name=None, history=None)
        wrapped_ctx = wrap_const(ctx_proxy)
        custom_forward_result = _interpret_call(custom_forward, wrapped_ctx, *args, **kwargs)
    
        # Forward.
        unwrapped_custom_forward_args = tree_map(lambda a: unwrap(a), args)
        unwrapped_custom_forward_result = unwrap(custom_forward_result)
        symbol_name = custom_autograd_function_cls.__name__
        custom_fwd_sym = Symbol(
            name=symbol_name,
            id=symbol_name,
            meta=lambda *unwrapped_custom_forward_args: unwrapped_custom_forward_result,
        )
        bsym_of_custom_fwd = BoundSymbol(
            custom_fwd_sym,
            args=unwrapped_custom_forward_args,
            kwargs={},
            output=unwrapped_custom_forward_result,
            subsymbols=custom_fwd_bsyms,
            source_filename=jit_ctx.computation_trace._current_source_filename,
            source_positions=None,
            _call_ctx=custom_fwd_bsyms[0]._call_ctx,
            _import_ctx=custom_fwd_bsyms[0]._import_ctx,
            _object_ctx=custom_fwd_bsyms[0]._object_ctx,
            _executor=custom_fwd_bsyms[0]._executor,
        )
        orig_scopes[-1].append(bsym_of_custom_fwd)
    
        augmented_bsym_output: tuple[tuple[TensorProxy, ...], tuple[TensorProxy, ...]] = (
            tuple(sequencify(unwrapped_custom_forward_result)),
            ctx_proxy.saved_tensors,
        )
        trace_of_augmented_fwd = TraceCtx()
        for bsym in custom_fwd_bsyms:
            trace_of_augmented_fwd.add_bound_symbol(bsym)
        trace_of_augmented_fwd.add_bound_symbol(prims.python_return.bind(augmented_bsym_output, output=()))
        si = SigInfo(custom_fwd_sym.name)
        for a in unwrapped_custom_forward_args:
            if isinstance(a, Proxy):
                si.args.append((a.name, None))
            else:
                pa = proxy(a)
                si.args.append((pa.name, None))
        trace_of_augmented_fwd._siginfo = si
        core_of_augmented_forward = trace_of_augmented_fwd.python_callable(include_decorators=False)
    
        @wraps(core_of_augmented_forward)
        def augmented_custom_forward_rule(*args, **kwargs):
            primal, residulas = core_of_augmented_forward(*args, **kwargs)
            check(len(primal) == 1, lambda f: "{primal=} has {len(primal)} proxies but expected 1")
            return VJPDual(primal=primal[0], residuals=residulas)
    
        augmented_forward_impls[custom_fwd_sym.name] = augmented_custom_forward_rule
    
        # Backward definition
        custom_backward = custom_autograd_function_cls.backward
        custom_bwd_bsyms: list[BoundSymbol] = []
        jit_ctx.computation_trace.scopes = [custom_bwd_bsyms]
        grads = tree_map(
            lambda a: TensorProxy(like=a).replace_name(f"grad_{a.name}"), sequencify(unwrapped_custom_forward_result)
        )
        wrapped_grads = tree_map(lambda g: wrap(g, provenance=custom_forward_result.provenance), grads)
        custom_backward_result = _interpret_call(custom_backward, wrapped_ctx, *wrapped_grads)
    
        trace_of_backward = TraceCtx()
        for bsym in custom_bwd_bsyms:
            trace_of_backward.add_bound_symbol(bsym)
>       trace_of_backward.add_bound_symbol(prims.python_return.bind(*unwrap(custom_backward_result), output=()))
E       TypeError: thunder.core.symbol.Symbol.bind() argument after * must be an iterable, not INTERPRETER_SIGNALS

thunder/core/jit_ext.py:684: TypeError

The above exception was the direct cause of the following exception:

requires_grad = False, memory_efficient = False

    @pytest.mark.parametrize("requires_grad", [True, False])
    @pytest.mark.parametrize("memory_efficient", [True, False])
    def test_apex_fused_rms_norm(requires_grad, memory_efficient):
        def fn(x, weight, normalized_shape, eps):
            return FusedRMSNormAffineMixedDtypesFunction.apply(x, weight, normalized_shape, eps, memory_efficient)
    
        device = "cuda"
        normalized_shape = (3, 2)
        x = torch.randn(4, 5, *normalized_shape, device=device, requires_grad=requires_grad)
        weight = torch.randn(*normalized_shape, device=device, requires_grad=requires_grad)
        eps = 1e-5
    
        expected = fn(x, weight, normalized_shape, eps)
        jfn = thunder.jit(fn, executors=[apex_ex])
>       actual = jfn(x, weight, normalized_shape, eps)

thunder/tests/test_apex_fused_norms.py:25: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
thunder/__init__.py:717: in fn_
    cache_entry, inps, pro_to_epi = get_computation_and_inputs(*args, **kwargs)
thunder/core/langctxs.py:136: in _fn
    result = fn(*args, **kwargs)
thunder/__init__.py:219: in cache_info_wrapper
    res = fn(*args, **kwargs)
thunder/__init__.py:506: in get_computation_and_inputs
    jit_results: TraceResults = thunder_general_jit(
thunder/core/jit_ext.py:1586: in thunder_general_jit
    result = jfn(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = <thunder.core.interpreter.WrappedValue object at 0x7507ec3db3a0>, kwargs = <thunder.core.interpreter.WrappedValue object at 0x7507ec3db370>, runtimectx = <thunder.core.interpreter.InterpreterRuntimeCtx object at 0x7507ec3db430>
fn_wrapped = <thunder.core.interpreter.WrappedValue object at 0x7507ec3db2b0>, getfn = <function interpret.<locals>.fn_.<locals>.getfn at 0x7507ec114a60>, wrapped_fn_2 = <thunder.core.interpreter.WrappedValue object at 0x7507ec3db280>
wrapped_closure = <thunder.core.interpreter.WrappedValue object at 0x7507ec3db1c0>, wrapped_cell = <thunder.core.interpreter.WrappedValue object at 0x7507ec3db130>, traceback_str = ''
msg = 'Encountered exception TypeError: thunder.core.symbol.Symbol.bind() argument after * must be an iterable, not INTERPRETER_SIGNALS while tracing <function test_apex_fused_rms_norm.<locals>.fn at 0x750828d3b6d0>:\n'

    @functools.wraps(fn)
    def fn_(*args, **kwargs) -> Any:
        runtimectx: InterpreterRuntimeCtx = InterpreterRuntimeCtx(debug_log=debug_log, record_history=record_history)
    
        with interpreter_ctx(compilectx, runtimectx):
            try:
                # we normalize the outmost function to be interpreted to take
                # args and kwargs as arguments (not *args and **kwargs).
                # We thus have three special INPUTs for the entry function: INPUT_ARGS, INPUT_KWARGS, INPUT_FN
                args = wrap(
                    args,
                    provenance=ProvenanceRecord(inst=PseudoInst.INPUT_ARGS, inputs=[]),
                )
    
                kwargs = wrap(
                    kwargs,
                    provenance=ProvenanceRecord(inst=PseudoInst.INPUT_KWARGS, inputs=[]),
                )
    
                fn_wrapped = wrap(
                    fn,
                    provenance=ProvenanceRecord(inst=PseudoInst.INPUT_FN, inputs=[]),
                )
    
                def getfn():
                    def fn_2(args, kwargs):
                        return fn(*args, **kwargs)
    
                    return fn_2
    
                wrapped_fn_2 = wrap_const(getfn())
                if compilectx._with_provenance_tracking:
                    wrapped_closure = wrap_attribute(
                        wrapped_fn_2.value.__closure__, wrapped_fn_2, wrap_const("__closure__")
                    )
                    wrapped_cell = wrap_binary_subscr(wrapped_closure.value[0], wrapped_closure, 0)
                    assert isinstance(wrapped_closure.item_wrappers, list)
                    wrapped_closure.item_wrappers[0] = wrapped_cell
                    populate_attribute_wrapper(wrapped_cell, "cell_contents", fn_wrapped)
    
                interpretation_result: Any = _interpret_call(wrapped_fn_2, args, kwargs)
                interpretation_result = unwrap(interpretation_result)
            except BaseException as e:
                # TODO Highlight the portion of the line that originated the opcode on Python versions that include
                #      the line offset information in the instruction
                traceback_str = os.linesep.join(f.format_with_source() for f in runtimectx.frame_stack)
                msg = (
                    f"Encountered exception {type(e).__name__}: {e} while tracing {fn}:{os.linesep}" f"{traceback_str}"
                )
>               raise InterpreterError(msg) from e
E               thunder.core.interpreter.InterpreterError: Encountered exception TypeError: thunder.core.symbol.Symbol.bind() argument after * must be an iterable, not INTERPRETER_SIGNALS while tracing <function test_apex_fused_rms_norm.<locals>.fn at 0x750828d3b6d0>:

thunder/core/interpreter.py:7120: InterpreterError
```

</details>

It looks like `_general_jit_torch_autograd_function_apply_lookaside` picks up the FusedRMSNorm.apply before it is picked up by the executor lookaside and fails inside `_general_jit_torch_autograd_function_apply_lookaside`.

https://github.com/Lightning-AI/lightning-thunder/blob/8d3561aae877f277bb820ce7cd74057399ccc7ac/thunder/core/jit_ext.py#L593-L605

@crcrpar do you think there would be a way to disable `_general_jit_torch_autograd_function_apply_lookaside` only for FusedRMSNorm?

```
thunder/tests/test_apex_fused_norms.py:25:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
thunder/__init__.py:717: in fn_
    cache_entry, inps, pro_to_epi = get_computation_and_inputs(*args, **kwargs)
thunder/core/langctxs.py:136: in _fn
    result = fn(*args, **kwargs)
thunder/__init__.py:219: in cache_info_wrapper
    res = fn(*args, **kwargs)
thunder/__init__.py:506: in get_computation_and_inputs
    jit_results: TraceResults = thunder_general_jit(
thunder/core/jit_ext.py:1593: in thunder_general_jit
    result = jfn(*args, **kwargs)
thunder/core/interpreter.py:7129: in fn_
    raise e
thunder/core/interpreter.py:7097: in fn_2
    return fn(*args, **kwargs)
thunder/tests/test_apex_fused_norms.py:15: in fn
    return FusedRMSNormAffineMixedDtypesFunction.apply(x, weight, normalized_shape, eps, memory_efficient)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    @staticmethod
    def backward(ctx, grad_output):
>       input_or_output, weight_, invvar = ctx.saved_tensors
E       ValueError: not enough values to unpack (expected 3, got 0)

../../../../.pyenv/versions/3.10.13/envs/torchdev-3.10/lib/python3.10/site-packages/apex-0.1-py3.10-linux-x86_64.egg/apex/normalization/fused_layer_norm.py:85: ValueError
```

--- a/thunder/core/jit_ext.py
+++ b/thunder/core/jit_ext.py
@@ -660,7 +660,13 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
             pa = proxy(a)
             si.args.append((pa.name, None))
     trace_of_augmented_fwd._siginfo = si
-    core_of_augmented_forward = trace_of_augmented_fwd.python_callable(include_decorators=False)
+    trace_of_augmented_fwd.args = unwrapped_custom_forward_args
+
+    @wraps(trace_of_augmented_fwd.python_callable())
+    def core_of_augmented_forward(*args, **kwargs):
+        return thunder.core.trace_interpreter.interpret_trace(trace_of_augmented_fwd, *args, **kwargs)
+
+    # core_of_augmented_forward = trace_of_augmented_fwd.python_callable(include_decorators=False)
 
     @wraps(core_of_augmented_forward)
     def augmented_custom_forward_rule(*args, **kwargs):
@@ -695,7 +701,13 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
             pa = proxy(a)
             bwd_si.args.append((pa.name, None))
     trace_of_backward._siginfo = bwd_si
-    bwd_trace_callable_interface = trace_of_backward.python_callable(include_decorators=False)
+    trace_of_backward.args = tuple(ctx_proxy.saved_tensors + grads)
+
+    @wraps(trace_of_backward.python_callable())
+    def bwd_trace_callable_interface(*args, **kwargs):
+        return thunder.core.trace_interpreter.interpret_trace(trace_of_backward, *args, **kwargs)
+
+    # bwd_trace_callable_interface = trace_of_backward.python_callable(include_decorators=False)
 
     bwd_si = SigInfo("backward_impl")
     for a in ctx_proxy.saved_consts + ctx_proxy.saved_tensors + grads:
@@ -709,7 +721,13 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
         bwd_trace_impl.add_bound_symbol(bsym)
     bwd_trace_impl.add_bound_symbol(prims.python_return.bind(*sequencify(unwrap(custom_backward_result)), output=()))
     bwd_trace_impl._siginfo = bwd_si
-    bwd_impl_callable = bwd_trace_impl.python_callable(include_decorators=False)
+    bwd_trace_impl.args = tuple(ctx_proxy.saved_consts + ctx_proxy.saved_tensors + grads)
+
+    @wraps(bwd_trace_impl.python_callable())
+    def bwd_impl_callable(*args, **kwargs):
+        return thunder.core.trace_interpreter.interpret_trace(bwd_trace_impl, *args, **kwargs)
+
+    # bwd_impl_callable = bwd_trace_impl.python_callable(include_decorators=False)
 
     @wraps(bwd_trace_callable_interface)
     def backward_impl(*args, **kwargs):

--- a/thunder/core/jit_ext.py
+++ b/thunder/core/jit_ext.py
@@ -644,6 +644,31 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
     )
     orig_scopes[-1].append(bsym_of_custom_fwd)
 
+    # not augmented for when we don't need grad
+    trace_of_fwd = TraceCtx()
+    for bsym in custom_fwd_bsyms:
+        trace_of_fwd.add_bound_symbol(bsym)
+    with tracectx(trace_of_fwd):
+        prims.python_return(unwrapped_custom_forward_result)
+
+    si = SigInfo(custom_fwd_sym.name)
+    for a in unwrapped_custom_forward_args:
+        if isinstance(a, Proxy):
+            si.args.append((a.name, None))
+        else:
+            pa = proxy(a)
+            si.args.append((pa.name, None))
+    trace_of_fwd._siginfo = si
+    trace_of_fwd.args = unwrapped_custom_forward_args
+
+    @wraps(trace_of_fwd.python_callable())
+    def core_of_forward(*args, **kwargs):
+        return thunder.core.trace_interpreter.interpret_trace(trace_of_fwd, *args, **kwargs)
+
+    thunder.executors.torchex._register_implementation(
+        custom_fwd_sym, core_of_forward, checker=thunder.executors.torchex._always_executable
+    )
+
     augmented_bsym_output: tuple[tuple[TensorProxy, ...], tuple[TensorProxy, ...]] = (
         tuple(sequencify(unwrapped_custom_forward_result)),
         ctx_proxy.saved_tensors,
@@ -707,8 +732,6 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
     def bwd_trace_callable_interface(*args, **kwargs):
         return thunder.core.trace_interpreter.interpret_trace(trace_of_backward, *args, **kwargs)
 
-    # bwd_trace_callable_interface = trace_of_backward.python_callable(include_decorators=False)
-
     bwd_si = SigInfo("backward_impl")
     for a in ctx_proxy.saved_consts + ctx_proxy.saved_tensors + grads:
         if isinstance(a, Proxy):
@@ -726,8 +749,6 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
     @wraps(bwd_trace_impl.python_callable())
     def bwd_impl_callable(*args, **kwargs):
         return thunder.core.trace_interpreter.interpret_trace(bwd_trace_impl, *args, **kwargs)
-
-    # bwd_impl_callable = bwd_trace_impl.python_callable(include_decorators=False)
 
     @wraps(bwd_trace_callable_interface)
     def backward_impl(*args, **kwargs):

--- a/thunder/executors/apex_fused_rms_norm_impl.py
+++ b/thunder/executors/apex_fused_rms_norm_impl.py
@@ -34,22 +34,6 @@ def apex_fused_rms_norm_forward_affine_meta(
     return TensorProxy(like=input), invvar
 
 
-apex_fused_rms_norm_forward_affine = apex_ex.register_operator(
-    "apex_fused_rms_norm_forward_affine",
-    meta=apex_fused_rms_norm_forward_affine_meta,
-    fn=fused_layer_norm_cuda.rms_forward_affine,
-    replaces=fused_layer_norm_cuda.rms_forward_affine,
-)
-
-
-apex_fused_rms_norm_forward_affine_mixed_dtypes = apex_ex.register_operator(
-    "apex_fused_rms_norm_forward_affine_mixed_dtypes",
-    meta=apex_fused_rms_norm_forward_affine_meta,
-    fn=fused_layer_norm_cuda.rms_forward_affine_mixed_dtypes,
-    replaces=fused_layer_norm_cuda.rms_forward_affine_mixed_dtypes,
-)
-
-
 def apex_fused_rms_norm_backward_affine_meta(
     grad_output: TensorLike,
     invvar: TensorLike,
@@ -62,105 +46,25 @@ def apex_fused_rms_norm_backward_affine_meta(
     return TensorProxy(like=grad_output), TensorProxy(like=weight_)
 
 
-apex_fused_rms_norm_backward_affine = apex_ex.register_operator(
-    "apex_fused_rms_norm_backward_affine",
-    meta=apex_fused_rms_norm_backward_affine_meta,
-    fn=fused_layer_norm_cuda.rms_backward_affine,
-    replaces=fused_layer_norm_cuda.rms_backward_affine,
-)
-
-
-def meta_impl_fn(
-    input: TensorLike, weight: TensorLike, normalized_shape: Sequence[int], eps: float, memory_efficient: bool
-):
-    output_or_input = TensorProxy(like=input)
-    weight = TensorProxy(like=input, shape=normalized_shape)
-    unnormalized_dims = len(input.shape) - len(normalized_shape)
-    invvar = TensorProxy(like=input, shape=(math.prod(input.shape[:unnormalized_dims]),))
-    return TensorProxy(like=input), (output_or_input, weight, invvar), AnyProxy(object())
-
-
-def fused_rms_norm_impl(
-    input: TensorLike, weight: TensorLike, normalized_shape: Sequence[int], eps: float, memory_efficient: bool
-):
-    ctx = Context()
-    output = FusedRMSNormAffineMixedDtypesFunction.forward(ctx, input, weight, normalized_shape, eps, memory_efficient)
-    return output, ctx.pop_saved_tensors(), ctx
-
-
-# fused_rms_norm_fwd = apex_ex.register_operator("fused_rms_norm_fwd", meta=meta_impl_fn, fn=fused_rms_norm_impl)
-
-
-def fused_rms_norm_backward_meta(saved_tensors: Sequence[torch.Tensor], ctx: Context, g: TensorLike):
-    # saved_tensors[0] - input or output
-    # saved_tensors[1] - weight
-    return TensorProxy(like=saved_tensors[0]), TensorProxy(like=saved_tensors[1])
-
-
-def fused_rms_norm_backward_impl(saved_tensors: Sequence[torch.Tensor], ctx: Context, g: TensorLike):
-    with set_saved_tensors(ctx, saved_tensors):
-        return FusedRMSNormAffineMixedDtypesFunction.backward(ctx, g)[:2]
-
-
-# fused_rms_norm_backward = apex_ex.register_operator(
-#    "fused_rms_norm_backward", meta=fused_rms_norm_backward_meta, fn=fused_rms_norm_backward_impl
-# )
-
-
-def fused_rms_norm_grad_rule(
-    ctx, input: TensorLike, weight: TensorLike, normalized_shape: Sequence[int], eps: float, memory_efficient: bool
-):
-    output, saved_tensors, saved_meta = fused_rms_norm_fwd(input, weight, normalized_shape, eps, memory_efficient)
-    g = get_grad(output)
-    grad_input, grad_weight = fused_rms_norm_backward(saved_tensors, saved_meta, g)
-    put_grads((input, weight), (grad_input, grad_weight))
-    return output
-
-
-def execution_tfms(
-    ctx, input: TensorLike, weight: TensorLike, normalized_shape: Sequence[int], eps: float, memory_efficient: bool
-):
-    output, _, _ = fused_rms_norm_fwd(input, weight, normalized_shape, eps, memory_efficient)
-    return output
-
-
-def _fused_rms_norm_checker(
-    ctx: torch.autograd.Function,
-    input: TensorLike,
-    weight: TensorLike,
-    normalized_shape: Sequence[int],
-    eps: float,
-    memory_efficient: bool,
-):
-    use_apex_fused_rms_norm = get_compile_option(
-        "use_apex_fused_rms_norm", "Whether to enable `fused_rms_norm` from `apex_ex`. Defaults to `True`."
-    )
-    # We explicitly check for `False` as if the value is unspecified by user, `get_compile_option` returns `None` and `not None` is equal to True.
-    if use_apex_fused_rms_norm == False:  # User explicitly disabled this.
-        return False
-
-    # use_apex_fused_rms_norm is `None` or `True`.
-    return True
-
-
 # Create a new symbol and register lookaside only if import is available.
 if apex_fused_norms_available():
-
-    def meta_fn(
-        ctx, input: TensorLike, weight: TensorLike, normalized_shape: Sequence[int], eps: float, memory_efficient: bool
-    ):
-        return TensorProxy(like=input)
-
-    # Symbol which will be used by lookaside.
-    fused_rms_norm = apex_ex.register_operator(
-        "fused_rms_norm",
-        meta=meta_fn,  # replaces=FusedRMSNormAffineMixedDtypesFunction.forward
+    apex_fused_rms_norm_forward_affine = apex_ex.register_operator(
+        "apex_fused_rms_norm_forward_affine",
+        meta=apex_fused_rms_norm_forward_affine_meta,
+        fn=fused_layer_norm_cuda.rms_forward_affine,
+        replaces=fused_layer_norm_cuda.rms_forward_affine,
     )
 
-    # apex_ex.register_implementation(
-    #    fused_rms_norm,
-    #    execution_transform=execution_tfms,
-    #    grad_transform=fused_rms_norm_grad_rule,
-    #    checker=_fused_rms_norm_checker,
-    # )
-    # apex_ex.register_implementation(fused_rms_norm_backward, fused_rms_norm_backward)
+    apex_fused_rms_norm_forward_affine_mixed_dtypes = apex_ex.register_operator(
+        "apex_fused_rms_norm_forward_affine_mixed_dtypes",
+        meta=apex_fused_rms_norm_forward_affine_meta,
+        fn=fused_layer_norm_cuda.rms_forward_affine_mixed_dtypes,
+        replaces=fused_layer_norm_cuda.rms_forward_affine_mixed_dtypes,
+    )
+
+    apex_fused_rms_norm_backward_affine = apex_ex.register_operator(
+        "apex_fused_rms_norm_backward_affine",
+        meta=apex_fused_rms_norm_backward_affine_meta,
+        fn=fused_layer_norm_cuda.rms_backward_affine,
+        replaces=fused_layer_norm_cuda.rms_backward_affine,
+    )

--- a/thunder/executors/apex_fused_rms_norm_impl.py
+++ b/thunder/executors/apex_fused_rms_norm_impl.py
@@ -24,6 +24,52 @@ def apex_fused_norms_available() -> bool:
     return APEX_FUSED_NORMS_AVAILABLE
 
 
+def apex_fused_rms_norm_forward_affine_meta(
+    input: TensorLike, normalized_shape: Sequence[int], weight: TensorLike, eps: float
+):
+    output_or_input = TensorProxy(like=input)
+    weight = TensorProxy(like=input, shape=normalized_shape)
+    unnormalized_dims = len(input.shape) - len(normalized_shape)
+    invvar = TensorProxy(like=input, shape=(math.prod(input.shape[:unnormalized_dims]),))
+    return TensorProxy(like=input), invvar
+
+
+apex_fused_rms_norm_forward_affine = apex_ex.register_operator(
+    "apex_fused_rms_norm_forward_affine",
+    meta=apex_fused_rms_norm_forward_affine_meta,
+    fn=fused_layer_norm_cuda.rms_forward_affine,
+    replaces=fused_layer_norm_cuda.rms_forward_affine,
+)
+
+
+apex_fused_rms_norm_forward_affine_mixed_dtypes = apex_ex.register_operator(
+    "apex_fused_rms_norm_forward_affine_mixed_dtypes",
+    meta=apex_fused_rms_norm_forward_affine_meta,
+    fn=fused_layer_norm_cuda.rms_forward_affine_mixed_dtypes,
+    replaces=fused_layer_norm_cuda.rms_forward_affine_mixed_dtypes,
+)
+
+
+def apex_fused_rms_norm_backward_affine_meta(
+    grad_output: TensorLike,
+    invvar: TensorLike,
+    input_or_output: TensorLike,
+    normalized_shape: Sequence[int],
+    weight_,
+    eps: float,
+    memory_efficient: bool,
+):
+    return TensorProxy(like=grad_output), TensorProxy(like=weight_)
+
+
+apex_fused_rms_norm_backward_affine = apex_ex.register_operator(
+    "apex_fused_rms_norm_backward_affine",
+    meta=apex_fused_rms_norm_backward_affine_meta,
+    fn=fused_layer_norm_cuda.rms_backward_affine,
+    replaces=fused_layer_norm_cuda.rms_backward_affine,
+)
+
+
 def meta_impl_fn(
     input: TensorLike, weight: TensorLike, normalized_shape: Sequence[int], eps: float, memory_efficient: bool
 ):
@@ -42,7 +88,7 @@ def fused_rms_norm_impl(
     return output, ctx.pop_saved_tensors(), ctx
 
 
-fused_rms_norm_fwd = apex_ex.register_operator("fused_rms_norm_fwd", meta=meta_impl_fn, fn=fused_rms_norm_impl)
+# fused_rms_norm_fwd = apex_ex.register_operator("fused_rms_norm_fwd", meta=meta_impl_fn, fn=fused_rms_norm_impl)
 
 
 def fused_rms_norm_backward_meta(saved_tensors: Sequence[torch.Tensor], ctx: Context, g: TensorLike):
@@ -56,9 +102,9 @@ def fused_rms_norm_backward_impl(saved_tensors: Sequence[torch.Tensor], ctx: Con
         return FusedRMSNormAffineMixedDtypesFunction.backward(ctx, g)[:2]
 
 
-fused_rms_norm_backward = apex_ex.register_operator(
-    "fused_rms_norm_backward", meta=fused_rms_norm_backward_meta, fn=fused_rms_norm_backward_impl
-)
+# fused_rms_norm_backward = apex_ex.register_operator(
+#    "fused_rms_norm_backward", meta=fused_rms_norm_backward_meta, fn=fused_rms_norm_backward_impl
+# )
 
 
 def fused_rms_norm_grad_rule(
@@ -107,13 +153,14 @@ if apex_fused_norms_available():
 
     # Symbol which will be used by lookaside.
     fused_rms_norm = apex_ex.register_operator(
-        "fused_rms_norm", meta=meta_fn, replaces=FusedRMSNormAffineMixedDtypesFunction.forward
+        "fused_rms_norm",
+        meta=meta_fn,  # replaces=FusedRMSNormAffineMixedDtypesFunction.forward
     )
 
-    apex_ex.register_implementation(
-        fused_rms_norm,
-        execution_transform=execution_tfms,
-        grad_transform=fused_rms_norm_grad_rule,
-        checker=_fused_rms_norm_checker,
-    )
-    apex_ex.register_implementation(fused_rms_norm_backward, fused_rms_norm_backward)
+    # apex_ex.register_implementation(
+    #    fused_rms_norm,
+    #    execution_transform=execution_tfms,
+    #    grad_transform=fused_rms_norm_grad_rule,
+    #    checker=_fused_rms_norm_checker,
+    # )
+    # apex_ex.register_implementation(fused_rms_norm_backward, fused_rms_norm_backward)

